### ISSUE TITLE
Returning self as dict value does validate against return type

### ISSUE THREAD

**Issue by gjcarneiro** [astral-sh/ty#1436]
### Summary

https://play.ty.dev/fcecf4ca-df7b-4dd5-8922-97219ceb68cf

```py

from __future__ import annotations

class Account:

    def get_account(self) -> dict[str, Account]:
        return dict(loan=self)
```

```
$ mypy /tmp/xxx.py 
Success: no issues found in 1 source file
$ ty check /tmp/xxx.py 
Checking ------------------------------------------------------------ 1/1 files                                                                                                                                                                                          
/tmp/xxx.py:6:16: error[invalid-return-type] 
Return type does not match returned value: expected `dict[str, Account]`, found `dict[str, Self@get_account]`
Found 1 diagnostic
```

**Comment by ibraheemdev** [comment]
I believe we need to special case the `dict` constructor for this to work.
